README.md
# Jefferson-Papers
Repo for my work on the Jefferson Papers Research Project

Steps to merge the jeff and loc corpora:

1. run `python scraper.py` with a solid internet connection if you need to download the corpus off the library of congress

2. run `python jeff_cleanup.py` provided you have the jeff corpus in your directory

3. run `python xmlConvert.py "Jefferson Papers"`

4. run `python separator.py`

5. run `python merge.py`

6. If you want a csv file of the filename, the writer of the letter, recipient of the letter, date (YYYY-MM-DD) and the corpus, run `python csvConvert.py`

The complete corpora will be under the "Jefferson Papers" directory.


TODO:

1. Delete "jeff" directory after user runs `merge.py`

2. Merge both relevant Jefferson Papers subdirectories and delete the XML files

changes.txt
george washington from thomas jefferson, january 15, 1792, with copy.txt -> to george washington, january 15, 1792.txt
john mason from thomas jefferson, august 18, 1814.txt -> to john mason, august 18, 1814.txt
william h. cabell from thomas jefferson, june 29, 1807.txt -> to william h. cabell, june 29, 1807.txt
v4.41.letter xli.--m. le comte diodati, march 29, 1807.txt -> to m. le comte diodati, march 29, 1807.txt
v3.205.letter cciii.--to james madison, january 1, 1797.txt -> to james madison, january 1, 1797.txt
v4.160.letter clx.--to --------- nicholas, december 11,1821.txt -> to --------- nicholas, december 11,1821.txt
v3.166.letter clxiv.--circular to the merchants of the u.s., august 23, 1793.txt -> to the merchants of the u.s., august 23, 1793.txt
v3.156.letters.--re the lost million, june 10, 1793.txt -> to the treasury department, june 10, 1793.txt
featured to john wayles eppes, april 1, 1816.txt -> to john wayles eppes, april 1, 1816.txt
v3.144.letter cxliii.--circular to morris, pinckney, and short, april 26,1793.txt -> to morris, pinckney, and short, april 26,1793.txt
v3.131.letter cxxx.--circular to the ministers, february 13, 1793.txt -> to the ministers, february 13, 1793.txt
v3.110.letter cix.--circular to the american consuls, may 31, 1792.txt -> to the american consuls, may 31, 1792.txt
v3.41.letter xli.--circular to the consuls, august 26, 1790.txt -> to the american consuls, august 26, 1790.txt
v2.memoranda taken on a journey from paris in 1787.txt -> memoranda taken on a journey from paris in 1787.txt
v2.29.letter xxx.--m. le roy de l'academie des sciences, november 13, 1786.txt -> to m. le roy de l'academie des sciences, november 13, 1786.txt
v1.questions for economie politique et diplomatique.txt -> to economie politique et diplomatique.txt
v1.118.letter cxviii.--james madison, october 2, 1785.txt -> to james madison, october 2, 1785.txt
v1.107.letter cvii.--lister asquith, september 14, 1785.txt -> to lister asquith, september 14, 1785.txt
v1.95.letter xcv.--to peter carr--advice to a young man, aug. 19, 1785.txt -> to peter carr, aug. 19, 1785.txt
v1.0.memoir.txt -> memoir.txt
to doctor benjamin rush.txt -> to doctor benjamin rush, december 5, 1811.txt
to john adams, to john adams.txt -> to john adams, january 11, 1817.txt
to john adams.txt -> to john adams, june 27, 1822.txt
to elbridge gerry.txt -> to elbridge gerry, january 26, 1799.txt
to the count de montmorin,.txt -> to the count de montmorin, september 11, 1788.txt
to william carmichael.txt -> to william carmichael, may 5, 1786.txt
to john page, august 20 1785.txt -> to john page, august 20, 1785.txt
to gouverneur morris,.txt -> to gouverneur morris, march 10, 1792.txt
to samuel adams.txt -> to samuel adams, february 26, 1800.txt
to a. donald, february.7, 1788.txt -> to a. donald, february 7, 1788.txt
to james madison.txt -> to james madison, december 17, 1796.txt
to the emperor of russia.txt -> to the emperor of russia, april 19, 1806.txt
benjmain h. latrobe to thomas jefferson, august 12, 1817, with thomas jefferson note.txt -> benjamin h. latrobe to thomas jefferson, august 12, 1817, with thomas jefferson note.txt
james monroe to thomas jefferson, december 1823, dated received december 11.txt -> james monroe to thomas jefferson, december 11, 1823.txt

















csvConvert.py
import csv
import os
import sys
import dateutil.parser as dparser
import re
from datetime import datetime



def main():
    with open('jeff_data.csv', 'w') as f:
        writer = csv.writer(f)
        writer.writerow(['Filename', 'From', 'To', 'Date', 'Randolph/LoC'])
        f.close()
    
    write("Jefferson Papers/From_Jefferson")
    write("Jefferson Papers/To_Jefferson")
            
        

def write(directory):
    files = os.listdir(directory)
    for file in files:
        source = "LoC"
        if file[0:2] == "R.":
            source = "Randolph"
            file = file[3: ]
        x = re.search("(jan(uary)?|feb(ruary)?|mar(ch)?|apr(il)?|may|jun(e)?|jul(y)?|aug(ust)?|sep(tember)?|oct(ober)?|nov(ember)?|dec(ember)?)(.)?\s+\d{1,2},(\s+\d{4}|\d{4})", file)
        if x is not None:
            time = file[x.start():x.end()]
            dt_object = dparser.parse(time).strftime('%Y-%m-%d')

            if directory == "Jefferson Papers/From_Jefferson":
                start = file.find("to")+3
                end = x.start()-2

                recipient = file[start: end]
                with open('jeff_data.csv', 'a') as f:
                    writer = csv.writer(f)
                    writer.writerow([file, "thomas jefferson", recipient, dt_object, source])
                    f.close()

            else:
                composer = file[0:file.find("to")-1]

                with open('jeff_data.csv', 'a') as f:
                    writer = csv.writer(f)
                    writer.writerow([file, composer, "thomas jefferson", dt_object, source])
                    f.close()



        else:
            if file == "thomas jefferson to uriah tracy, january 1806.txt":
                with open('jeff_data.csv', 'a') as f:
                    writer = csv.writer(f)
                    writer.writerow([file, "thomas jefferson", "uriah tracy", "1806-01-XX", source])
                    f.close()
                continue

            elif file == "memoir.txt":
                with open('jeff_data.csv', 'a') as f:
                    writer = csv.writer(f)
                    writer.writerow([file, "thomas jefferson", "", "XXXX-XX-XX", source])
                    f.close()
                continue

            elif file == "to colonel mathews, october, 1779.txt":
                with open('jeff_data.csv', 'a') as f:
                    writer = csv.writer(f)
                    writer.writerow([file, "thomas jefferson", "colonel mathews", "1779-10-XX", source])
                    f.close()
                continue

            elif file == "to dr. gem 1792.txt":
                with open('jeff_data.csv', 'a') as f:
                    writer = csv.writer(f)
                    writer.writerow([file, "thomas jefferson", "dr. gem", "1792-XX-XX", source])
                    f.close()
                continue

            elif file == "memoranda taken on a journey from paris in 1787.txt":
                with open('jeff_data.csv', 'a') as f:
                    writer = csv.writer(f)
                    writer.writerow([file, "thomas jefferson", "", "1787-XX-XX", source])
                    f.close()
                continue

            elif file == "to economie politique et diplomatique.txt":
                with open('jeff_data.csv', 'a') as f:
                    writer = csv.writer(f)
                    writer.writerow([file, "thomas jefferson", "economie politique et diplomatique", "XXXX-XX-XX", source])
                    f.close()
                continue

            elif file == "thomas jefferson, march 1809, statement on appointments.txt":
                with open('jeff_data.csv', 'a') as f:
                    writer = csv.writer(f)
                    writer.writerow([file, "thomas jefferson", "", "1809-03-XX", source])
                    f.close()
                continue

            elif file == "james monroe to thomas jefferson, november 1797.txt":
                with open('jeff_data.csv', 'a') as f:
                    writer = csv.writer(f)
                    writer.writerow([file, "james monroe", "thomas jefferson", "1797-11-XX", source])
                    f.close()
                continue

            elif file == ".DS_Store":
                continue

            elif file == "george washington to madame de lafayette, march 15, 1793.txt":
                continue
            
            elif file == "henry dearborn to state governors, january 17, 1809, circular letter.txt":
                continue

            elif file == "senate, february 1, 1791, resolution on redemption of american citizens held by algiers.txt":
                continue
            
            elif file == "cabinet to george washington, august 5, 1793, report.txt":
                continue

            elif file == "from the president, april 4, 1791.txt":
                continue
            
            print(file)

    





if __name__ == '__main__':
    main()